File: src/gpx.js
const EARTH_RADIUS_KM = 6371;
export const SPEED_WINDOW_SEC = 10;
export const POWER_WINDOW_SEC = 3;
export const CADENCE_WINDOW_SEC = 3;

function haversineKm(lat1, lon1, lat2, lon2) {
  const toRad = (deg) => (deg * Math.PI) / 180;
  const dLat = toRad(lat2 - lat1);
  const dLon = toRad(lon2 - lon1);
  const a =
    Math.sin(dLat / 2) ** 2 +
    Math.cos(toRad(lat1)) * Math.cos(toRad(lat2)) * Math.sin(dLon / 2) ** 2;
  return 2 * EARTH_RADIUS_KM * Math.asin(Math.sqrt(a));
}

function parseTime(text) {
  const d = new Date(text.trim());
  if (Number.isNaN(d.getTime())) {
    throw new Error(`Некорректное время в GPX: ${text}`);
  }
  return d;
}

function parseOptionalNumber(text) {
  if (text == null || text === "") return null;
  const n = parseFloat(text);
  return Number.isNaN(n) ? null : n;
}

function parseExtensions(pt) {
  const result = { hr: null, power: null, cadence: null, atemp: null };
  const ext = pt.querySelector("extensions");
  if (!ext) return result;

  for (const el of ext.getElementsByTagName("*")) {
    const name = el.localName?.toLowerCase();
    const value = parseOptionalNumber(el.textContent);
    if (value == null) continue;

    if (name === "hr") result.hr = value;
    else if (name === "cad") result.cadence = value;
    else if (name === "power") result.power = value;
    else if (name === "atemp") result.atemp = value;
  }

  return result;
}

function indexAtOrAfter(points, timeMs) {
  let lo = 0;
  let hi = points.length - 1;
  while (lo < hi) {
    const mid = Math.floor((lo + hi) / 2);
    if (points[mid].time.getTime() < timeMs) lo = mid + 1;
    else hi = mid;
  }
  return lo;
}

function indexAtOrBefore(points, timeMs) {
  let lo = 0;
  let hi = points.length - 1;
  while (lo < hi) {
    const mid = Math.ceil((lo + hi) / 2);
    if (points[mid].time.getTime() > timeMs) hi = mid - 1;
    else lo = mid;
  }
  return lo;
}

function getWindowRange(raw, i, windowSec) {
  const n = raw.length;
  const halfMs = (windowSec * 1000) / 2;
  const trackStart = raw[0].time.getTime();
  const trackEnd = raw[n - 1].time.getTime();
  const centerMs = raw[i].time.getTime();
  const startMs = Math.max(trackStart, centerMs - halfMs);
  const endMs = Math.min(trackEnd, centerMs + halfMs);

  let j = indexAtOrAfter(raw, startMs);
  let k = indexAtOrBefore(raw, endMs);

  if (j > k) {
    const tmp = j;
    j = k;
    k = tmp;
  }

  return { j, k };
}

/** power === 0 или отсутствует (null) → cadence = 0, до сглаживания */
function applyCadenceZeroWhenNoPower(raw) {
  for (const pt of raw) {
    if (pt.power === 0 || pt.power === null) {
      pt.cadence = 0;
    }
  }
}

function computeSmoothedSpeeds(raw) {
  const n = raw.length;
  if (n === 0) return [];
  if (n === 1) return [0];

  const speeds = [];

  for (let i = 0; i < n; i++) {
    let { j, k } = getWindowRange(raw, i, SPEED_WINDOW_SEC);

    if (j === k) {
      j = Math.max(0, i - 1);
      k = Math.min(n - 1, i + 1);
    }

    const dtMs = raw[k].time.getTime() - raw[j].time.getTime();
    if (dtMs <= 0) {
      speeds.push(0);
      continue;
    }

    const distKm = haversineKm(raw[j].lat, raw[j].lon, raw[k].lat, raw[k].lon);
    speeds.push((distKm / dtMs) * 3_600_000);
  }

  return speeds;
}

/**
 * Скользящее среднее по симметричному окну ±windowSec/2.
 * null в selector — точка пропускается; 0 учитывается.
 */
function computeWindowAverage(raw, windowSec, selector) {
  const n = raw.length;
  if (n === 0) return [];

  const values = [];

  for (let i = 0; i < n; i++) {
    const { j, k } = getWindowRange(raw, i, windowSec);
    let sum = 0;
    let count = 0;

    for (let idx = j; idx <= k; idx++) {
      const value = selector(raw[idx]);
      if (value == null) continue;
      sum += value;
      count++;
    }

    values.push(count > 0 ? sum / count : null);
  }

  return values;
}

/** Каденс: одностороннее окно [t − CADENCE_WINDOW_SEC, t], только idx <= i; при cadence=0 на i → 0. */
function computeSmoothedCadence(raw) {
  const n = raw.length;
  if (n === 0) return [];

  const windowMs = CADENCE_WINDOW_SEC * 1000;
  const values = [];

  for (let i = 0; i < n; i++) {
    if (raw[i].cadence === 0) {
      values.push(0);
      continue;
    }

    const centerMs = raw[i].time.getTime();
    const startMs = Math.max(raw[0].time.getTime(), centerMs - windowMs);
    let j = indexAtOrAfter(raw, startMs);
    if (j > i) j = i;

    let sum = 0;
    let count = 0;

    for (let idx = j; idx <= i; idx++) {
      const value = raw[idx].cadence;
      if (value == null) continue;
      sum += value;
      count++;
    }

    values.push(count > 0 ? sum / count : null);
  }

  return values;
}

/**
 * @param {File} file
 * @returns {Promise<{
 *   points: { time: Date, speed: number, hr: number|null, power: number|null, cadence: number|null, atemp: number|null }[],
 *   startTime: Date
 * }>}
 */
export function parseGPX(file) {
  return new Promise((resolve, reject) => {
    const reader = new FileReader();
    reader.onload = () => {
      try {
        const doc = new DOMParser().parseFromString(reader.result, "application/xml");
        if (doc.querySelector("parsererror")) {
          reject(new Error("Не удалось разобрать GPX файл"));
          return;
        }

        const trkpts = [...doc.querySelectorAll("trkpt")];
        const raw = [];

        for (const pt of trkpts) {
          const timeEl = pt.querySelector("time");
          if (!timeEl?.textContent?.trim()) continue;

          const lat = parseFloat(pt.getAttribute("lat"));
          const lon = parseFloat(pt.getAttribute("lon"));
          if (Number.isNaN(lat) || Number.isNaN(lon)) continue;

          const { hr, power, cadence, atemp } = parseExtensions(pt);
          raw.push({ lat, lon, time: parseTime(timeEl.textContent), hr, power, cadence, atemp });
        }

        if (raw.length === 0) {
          reject(new Error("GPX не содержит точек с тегом <time>"));
          return;
        }

        const rawLogFrom = new Date("2026-05-15T07:24:40Z");
        const rawLogTo = new Date("2026-05-15T07:25:00Z");
        for (const pt of raw) {
          if (pt.time >= rawLogFrom && pt.time <= rawLogTo) {
            console.log(
              "RAW:",
              pt.time.toISOString(),
              "power=",
              pt.power,
              "cadence=",
              pt.cadence,
            );
          }
        }

        let noPowerLogCount = 0;
        for (const pt of raw) {
          if (pt.power === null && noPowerLogCount < 3) {
            console.log("no power at", pt.time);
            noPowerLogCount++;
          }
        }

        applyCadenceZeroWhenNoPower(raw);

        const speeds = computeSmoothedSpeeds(raw);
        const powers = computeWindowAverage(raw, POWER_WINDOW_SEC, (pt) => pt.power);
        const cadences = computeSmoothedCadence(raw);

        const points = raw.map((pt, i) => ({
          time: pt.time,
          speed: speeds[i],
          hr: pt.hr,
          power: powers[i],
          cadence: cadences[i],
          atemp: pt.atemp,
        }));

        const smoothLogFrom = new Date("2026-05-15T07:24:47Z");
        const smoothLogTo = new Date("2026-05-15T07:24:52Z");
        for (const point of points) {
          if (point.time >= smoothLogFrom && point.time <= smoothLogTo) {
            console.log(
              "SMOOTHED:",
              point.time.toISOString(),
              "cadence=",
              point.cadence,
              "power=",
              point.power,
            );
          }
        }

        console.log("GPX points sample:", points.slice(0, 5));
        resolve({ points, startTime: points[0].time });
      } catch (err) {
        reject(err);
      }
    };
    reader.onerror = () => reject(new Error("Ошибка чтения GPX файла"));
    reader.readAsText(file);
  });
}

File: server/app/gpx.py
from __future__ import annotations

import math
from dataclasses import dataclass
from datetime import datetime, timezone


EARTH_RADIUS_KM = 6371.0
SPEED_WINDOW_SEC = 10
POWER_WINDOW_SEC = 3
CADENCE_WINDOW_SEC = 3


@dataclass
class GpxPoint:
    time: datetime
    lat: float
    lon: float
    hr: float | None
    power: float | None
    cadence: float | None
    speed: float = 0.0


def haversine_km(lat1: float, lon1: float, lat2: float, lon2: float) -> float:
    to_rad = math.radians
    d_lat = to_rad(lat2 - lat1)
    d_lon = to_rad(lon2 - lon1)
    a = (
        math.sin(d_lat / 2) ** 2
        + math.cos(to_rad(lat1)) * math.cos(to_rad(lat2)) * math.sin(d_lon / 2) ** 2
    )
    return 2 * EARTH_RADIUS_KM * math.asin(math.sqrt(a))


def parse_iso_datetime(text: str) -> datetime:
    text = text.strip()
    if text.endswith("Z"):
        text = text[:-1] + "+00:00"
    dt = datetime.fromisoformat(text)
    if dt.tzinfo is None:
        dt = dt.replace(tzinfo=timezone.utc)
    return dt.astimezone(timezone.utc)


GPX_NS = "http://www.topografix.com/GPX/1/1"


def local_name(tag: str) -> str:
    return tag.rsplit("}", 1)[-1]


def find_extensions_element(trkpt) -> object | None:
    """Strava/Garmin GPX: <extensions> в default namespace GPX 1.1."""
    ext = trkpt.find(f"{{{GPX_NS}}}extensions")
    if ext is not None:
        return ext
    for child in trkpt:
        if local_name(child.tag) == "extensions":
            return child
    return None


def parse_optional_number(text: str | None) -> float | None:
    if text is None or not str(text).strip():
        return None
    try:
        return float(text)
    except ValueError:
        return None


def parse_extensions(trkpt) -> tuple[float | None, float | None, float | None]:
    hr = power = cadence = None
    ext = find_extensions_element(trkpt)
    if ext is None:
        return hr, power, cadence
    for el in ext.iter():
        if el is ext:
            continue
        name = local_name(el.tag).lower()
        value = parse_optional_number(el.text)
        if value is None:
            continue
        if name == "hr":
            hr = value
        elif name == "cad":
            cadence = value
        elif name == "power":
            power = value
    return hr, power, cadence


def index_at_or_after(points: list[GpxPoint], time_ms: float) -> int:
    lo, hi = 0, len(points) - 1
    while lo < hi:
        mid = (lo + hi) // 2
        if points[mid].time.timestamp() * 1000 < time_ms:
            lo = mid + 1
        else:
            hi = mid
    return lo


def index_at_or_before(points: list[GpxPoint], time_ms: float) -> int:
    lo, hi = 0, len(points) - 1
    while lo < hi:
        mid = (lo + hi + 1) // 2
        if points[mid].time.timestamp() * 1000 > time_ms:
            hi = mid - 1
        else:
            lo = mid
    return lo


def get_window_range(points: list[GpxPoint], i: int, window_sec: int) -> tuple[int, int]:
    half_ms = window_sec * 500
    track_start = points[0].time.timestamp() * 1000
    track_end = points[-1].time.timestamp() * 1000
    center_ms = points[i].time.timestamp() * 1000
    start_ms = max(track_start, center_ms - half_ms)
    end_ms = min(track_end, center_ms + half_ms)
    j = index_at_or_after(points, start_ms)
    k = index_at_or_before(points, end_ms)
    if j > k:
        j, k = k, j
    return j, k


def apply_cadence_zero_when_no_power(points: list[GpxPoint]) -> None:
    for pt in points:
        if pt.power == 0 or pt.power is None:
            pt.cadence = 0


def compute_smoothed_speeds(points: list[GpxPoint]) -> None:
    n = len(points)
    if n == 0:
        return
    if n == 1:
        points[0].speed = 0
        return
    for i, pt in enumerate(points):
        j, k = get_window_range(points, i, SPEED_WINDOW_SEC)
        if j == k:
            j = max(0, i - 1)
            k = min(n - 1, i + 1)
        dt_ms = (points[k].time - points[j].time).total_seconds() * 1000
        if dt_ms <= 0:
            pt.speed = 0
            continue
        dist = haversine_km(points[j].lat, points[j].lon, points[k].lat, points[k].lon)
        pt.speed = (dist / dt_ms) * 3_600_000


def compute_window_average(
    points: list[GpxPoint], window_sec: int, getter
) -> list[float | None]:
    values: list[float | None] = []
    for i in range(len(points)):
        j, k = get_window_range(points, i, window_sec)
        total = 0.0
        count = 0
        for idx in range(j, k + 1):
            value = getter(points[idx])
            if value is None:
                continue
            total += value
            count += 1
        values.append(total / count if count else None)
    return values


def compute_smoothed_cadence(points: list[GpxPoint]) -> list[float | None]:
    window_ms = CADENCE_WINDOW_SEC * 1000
    track_start_ms = points[0].time.timestamp() * 1000
    values: list[float | None] = []
    for i, pt in enumerate(points):
        if pt.cadence == 0:
            values.append(0)
            continue
        center_ms = pt.time.timestamp() * 1000
        start_ms = max(track_start_ms, center_ms - window_ms)
        j = index_at_or_after(points, start_ms)
        if j > i:
            j = i
        total = 0.0
        count = 0
        for idx in range(j, i + 1):
            value = points[idx].cadence
            if value is None:
                continue
            total += value
            count += 1
        values.append(total / count if count else None)
    return values


def parse_gpx(path) -> list[GpxPoint]:
    import xml.etree.ElementTree as ET

    tree = ET.parse(path)
    root = tree.getroot()
    raw: list[GpxPoint] = []

    for trkpt in root.iter():
        if local_name(trkpt.tag) != "trkpt":
            continue
        time_el = None
        for child in trkpt:
            if local_name(child.tag) == "time":
                time_el = child
                break
        if time_el is None or not (time_el.text or "").strip():
            continue
        lat = float(trkpt.attrib["lat"])
        lon = float(trkpt.attrib["lon"])
        hr, power, cadence = parse_extensions(trkpt)
        raw.append(
            GpxPoint(
                time=parse_iso_datetime(time_el.text),
                lat=lat,
                lon=lon,
                hr=hr,
                power=power,
                cadence=cadence,
            )
        )

    if not raw:
        raise ValueError("GPX не содержит точек с тегом <time>")

    apply_cadence_zero_when_no_power(raw)
    compute_smoothed_speeds(raw)
    powers = compute_window_average(raw, POWER_WINDOW_SEC, lambda p: p.power)
    cadences = compute_smoothed_cadence(raw)

    for i, pt in enumerate(raw):
        pt.power = powers[i]
        pt.cadence = cadences[i]

    return raw


def nearest_point(points: list[GpxPoint], target: datetime) -> GpxPoint | None:
    if not points:
        return None
    target_ms = target.timestamp() * 1000
    first_ms = points[0].time.timestamp() * 1000
    last_ms = points[-1].time.timestamp() * 1000
    if target_ms < first_ms or target_ms > last_ms:
        return None

    lo, hi = 0, len(points) - 1
    while lo < hi:
        mid = (lo + hi) // 2
        if points[mid].time.timestamp() * 1000 < target_ms:
            lo = mid + 1
        else:
            hi = mid

    if lo > 0:
        prev = points[lo - 1]
        curr = points[lo]
        prev_diff = abs(prev.time.timestamp() * 1000 - target_ms)
        curr_diff = abs(curr.time.timestamp() * 1000 - target_ms)
        return prev if prev_diff <= curr_diff else curr
    return points[lo]
